File: test/sdk-examples/python-example/runtime_helpers.py
import os
import time
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

try:
    from tokvera import (
        TokveraOTelSpanExporter,
        configure_claude_agent_sdk,
        configure_google_adk,
        create_crewai_tracer,
        create_instructor_tracer,
        create_langgraph_tracer,
    create_pydanticai_tracer,
    create_tracer,
    finish_span,
    get_track_kwargs_from_trace_context,
    start_span,
    start_trace,
    track_mistral,
    )
except ImportError as exc:  # pragma: no cover - explicit operator guidance
    raise RuntimeError(
        "runtime_helpers.py requires tokvera with runtime-helper support (0.2.8+) "
        "or a local ../tokvera-python checkout on PYTHONPATH."
    ) from exc


def load_local_env() -> None:
    env_path = Path(__file__).with_name(".env")
    if not env_path.exists():
        return

    for raw_line in env_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue

        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def next_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex}"


@dataclass
class FakeMistralUsage:
    prompt_tokens: int = 11
    completion_tokens: int = 9
    total_tokens: int = 20


@dataclass
class FakeMistralResponse:
    model: str = "mistral-small-latest"
    usage: FakeMistralUsage = field(default_factory=FakeMistralUsage)
    choices: list[dict[str, Any]] = field(
        default_factory=lambda: [{"message": {"content": "Drafted a concise answer."}}]
    )


class FakeMistralChat:
    def complete(self, *args: Any, **kwargs: Any) -> FakeMistralResponse:
        _ = args, kwargs
        return FakeMistralResponse()


class FakeMistralClient:
    def __init__(self) -> None:
        self.chat = FakeMistralChat()


def run_existing_app(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    root = start_trace(
        tracer,
        step_name="existing_app_router",
        model="custom-router",
        span_kind="orchestrator",
        quality_label="good",
        feedback_score=4.9,
        outcome="success",
    )
    policy = start_span(
        root,
        step_name="route_policy",
        provider="tokvera",
        model="policy-engine",
        span_kind="guardrail",
        routing_reason="default_policy",
        route="anthropic:claude-3.5-haiku",
        quality_label="good",
    )
    finish_span(
        policy,
        response={"selected_route": "anthropic:claude-3.5-haiku"},
        quality_label="good",
        outcome="success",
    )
    finish_span(
        root,
        response={"completed": True},
        quality_label="good",
        feedback_score=5,
        outcome="success",
    )


def run_mistral(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    root = start_trace(
        tracer,
        step_name="mistral_support_flow",
        model="mistral-router",
        span_kind="orchestrator",
        quality_label="poor",
        feedback_score=2.2,
        outcome="success",
    )
    client = track_mistral(
        FakeMistralClient(),
        **get_track_kwargs_from_trace_context(
            root,
            step_name="mistral_reply",
            span_kind="model",
            provider="mistral",
            model="mistral-small-latest",
            quality_label="poor",
            feedback_score=2.1,
            outcome="success",
            emit_lifecycle_events=True,
            capture_content=True,
        ),
    )
    response = client.chat.complete(
        model="mistral-small-latest",
        messages=[{"role": "user", "content": "Draft a short refund policy answer."}],
    )
    finish_span(
        root,
        response={"completed": True, "provider_response": response.choices[0]["message"]["content"]},
        quality_label="poor",
        feedback_score=2.2,
        outcome="success",
    )


def run_claude_agent(base_options: dict[str, Any], feature: str) -> None:
    tracer = configure_claude_agent_sdk(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    run = tracer.start_agent(
        step_name="claude_support_agent",
        model="claude-agent-router",
        quality_label="good",
        feedback_score=4.6,
        outcome="success",
    )
    tool = tracer.start_tool(run, step_name="crm_lookup", tool_name="crm_lookup")
    tracer.finish_tool(tool, response={"customer_tier": "priority"}, quality_label="good", outcome="success")
    model = tracer.start_model(
        run,
        step_name="draft_reply",
        provider="anthropic",
        model="claude-3.5-haiku",
        quality_label="good",
        feedback_score=4.8,
        outcome="success",
    )
    tracer.finish_model(
        model,
        response={"output_text": "Priority customers are eligible for faster review."},
        usage={"prompt_tokens": 12, "completion_tokens": 10, "total_tokens": 22},
        quality_label="good",
        feedback_score=4.9,
        outcome="success",
    )
    tracer.finish_run(run, response={"status": "completed"}, quality_label="good", feedback_score=5, outcome="success")


def run_google_adk(base_options: dict[str, Any], feature: str) -> None:
    tracer = configure_google_adk(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    run = tracer.start_agent(
        step_name="google_adk_agent",
        model="adk-router",
        quality_label="good",
        feedback_score=4.5,
        outcome="success",
    )
    model = tracer.start_model(
        run,
        step_name="grounded_answer",
        provider="gemini",
        model="gemini-2.0-flash",
        quality_label="good",
        feedback_score=4.7,
        outcome="success",
    )
    tracer.finish_model(
        model,
        response={"output_text": "The account is now verified and ready for refund review."},
        usage={"prompt_tokens": 10, "completion_tokens": 9, "total_tokens": 19},
        quality_label="good",
        feedback_score=4.8,
        outcome="success",
    )
    tracer.finish_run(run, response={"status": "completed"}, quality_label="good", feedback_score=4.8, outcome="success")


def run_langgraph(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_langgraph_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    graph = tracer.start_graph(
        step_name="langgraph_support_flow",
        model="planner-graph",
        quality_label="good",
        feedback_score=4.4,
        outcome="success",
    )
    planner = tracer.start_node(graph, step_name="planner")
    tracer.finish_node(planner, response={"next": "kb_lookup"}, quality_label="good", outcome="success")
    branch = tracer.start_branch(graph, step_name="confidence_gate", routing_reason="confidence_below_threshold", route="fallback_path")
    tracer.finish_branch(
        branch,
        response={"selected_branch": "fallback_path"},
        quality_label="good",
        feedback_score=4.6,
        outcome="success",
    )
    tracer.finish_run(graph, response={"status": "completed"}, quality_label="good", feedback_score=4.7, outcome="success")


def run_instructor(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_instructor_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    extraction = tracer.start_extraction(
        step_name="invoice_extract",
        model="instructor-router",
        quality_label="good",
        feedback_score=4.3,
        outcome="success",
    )
    validation = tracer.start_validation(
        extraction,
        step_name="validation_retry",
        retry_reason="schema_mismatch",
    )
    tracer.finish_branch(
        validation,
        response={"schema_valid": True},
        quality_label="good",
        feedback_score=4.5,
        outcome="success",
    )
    tracer.finish_run(extraction, response={"status": "completed"}, quality_label="good", feedback_score=4.6, outcome="success")


def run_pydanticai(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_pydanticai_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    run = tracer.start_agent(
        step_name="pydanticai_run",
        model="validation-agent",
        quality_label="good",
        feedback_score=4.4,
        outcome="success",
    )
    validation = tracer.start_validation(
        run,
        step_name="pydantic_validation",
        retry_reason="model_parse_retry",
    )
    tracer.finish_branch(
        validation,
        response={"validated": True},
        quality_label="good",
        feedback_score=4.5,
        outcome="success",
    )
    tracer.finish_run(run, response={"status": "completed"}, quality_label="good", feedback_score=4.6, outcome="success")


def run_crewai(base_options: dict[str, Any], feature: str) -> None:
    tracer = create_crewai_tracer(
        **base_options,
        feature=feature,
        emit_lifecycle_events=True,
        capture_content=True,
    )
    crew = tracer.start_crew(
        step_name="crew_support_workflow",
        model="crew-router",
        quality_label="good",
        feedback_score=4.2,
        outcome="success",
    )
    tool = tracer.start_tool(crew, step_name="kb_search", tool_name="kb_search")
    tracer.finish_tool(tool, response={"match_count": 3}, quality_label="good", outcome="success")
    tracer.finish_run(crew, response={"status": "completed"}, quality_label="good", feedback_score=4.4, outcome="success")


def run_otel(base_options: dict[str, Any], feature: str) -> None:
    exporter = TokveraOTelSpanExporter(
        **base_options,
        feature=feature,
    )

    class _Status:
        is_ok = True

    class _Context:
        trace_id = next_id("trc")
        span_id = next_id("spn")

    class _Span:
        name = "retrieval_plan"
        start_time = 100.0
        end_time = 100.18
        parent_span_id = None
        attributes = {
            "tokvera.provider": "openai",
            "tokvera.feature": feature,
            "tokvera.step_name": "retrieval_plan",
            "tokvera.endpoint": "responses.create",
            "tokvera.cost_usd": 0.00002,
            "gen_ai.response.model": "gpt-4o-mini",
            "gen_ai.usage.input_tokens": 8,
            "gen_ai.usage.output_tokens": 5,
        }
        resource = type("Resource", (), {"attributes": {"service.name": "vector-router", "deployment.environment": "test"}})()
        status = _Status()

        def get_span_context(self):
            return _Context()

    exporter.export([_Span()])


def main() -> None:
    load_local_env()

    api_key = os.getenv("TOKVERA_API_KEY")
    if not api_key:
        raise RuntimeError("TOKVERA_API_KEY is required")

    base_options = {
        "api_key": api_key,
        "tenant_id": "example_tenant",
        "customer_id": "example_customer",
        "environment": "test",
        "schema_version": "2026-04-01",
    }

    features = {
        "existing_app": os.getenv("TOKVERA_FEATURE_EXISTING_APP_PY", "runtime_existing_app_py"),
        "mistral": os.getenv("TOKVERA_FEATURE_MISTRAL_PY", "runtime_mistral_py"),
        "claude_agent": os.getenv("TOKVERA_FEATURE_CLAUDE_AGENT_PY", "runtime_claude_agent_py"),
        "google_adk": os.getenv("TOKVERA_FEATURE_GOOGLE_ADK_PY", "runtime_google_adk_py"),
        "langgraph": os.getenv("TOKVERA_FEATURE_LANGGRAPH_PY", "runtime_langgraph_py"),
        "instructor": os.getenv("TOKVERA_FEATURE_INSTRUCTOR_PY", "runtime_instructor_py"),
        "pydanticai": os.getenv("TOKVERA_FEATURE_PYDANTICAI_PY", "runtime_pydanticai_py"),
        "crewai": os.getenv("TOKVERA_FEATURE_CREWAI_PY", "runtime_crewai_py"),
        "otel": os.getenv("TOKVERA_FEATURE_OTEL_PY", "runtime_otel_py"),
    }

    run_existing_app(base_options, features["existing_app"])
    run_mistral(base_options, features["mistral"])
    run_claude_agent(base_options, features["claude_agent"])
    run_google_adk(base_options, features["google_adk"])
    run_langgraph(base_options, features["langgraph"])
    run_instructor(base_options, features["instructor"])
    run_pydanticai(base_options, features["pydanticai"])
    run_crewai(base_options, features["crewai"])
    run_otel(base_options, features["otel"])

    time.sleep(float(os.getenv("TOKVERA_WAIT_SECONDS", "4.0")))
    print("python runtime helper smoke complete")
    print(features)


if __name__ == "__main__":
    main()

File: test/sdk-examples/node-example/runtime-helpers.mjs
import crypto from "node:crypto";
import {
  TokveraOTelSpanExporter,
  createTokveraLangGraphHooks,
  createTokveraOpenAIAgentsTracingProcessor,
  createTokveraTracer,
  finishSpan,
  getTrackOptionsFromTraceContext,
  startSpan,
  startTrace,
  trackMistral,
} from "@tokvera/sdk";
import { loadLocalEnv } from "./env.mjs";

function nextId(prefix) {
  return `${prefix}_${crypto.randomUUID().replace(/-/g, "")}`;
}

function createFakeMistralClient() {
  return {
    chat: {
      async complete(payload) {
        return {
          id: "mistral_resp_1",
          model: payload.model || "mistral-small-latest",
          usage: {
            prompt_tokens: 13,
            completion_tokens: 11,
            total_tokens: 24,
          },
          choices: [
            {
              message: {
                content: "Drafted a short answer for the customer.",
              },
            },
          ],
        };
      },
    },
  };
}

async function runExistingAppFeature(baseOptions, feature) {
  const tracer = createTokveraTracer({
    ...baseOptions,
    feature,
    emit_lifecycle_events: true,
    capture_content: true,
  });

  const root = startTrace(tracer.baseOptions, {
    step_name: "existing_app_router",
    model: "custom-router",
    span_kind: "orchestrator",
    payload_blocks: [
      {
        payload_type: "context",
        content: "Support router policy: prefer direct answer unless escalation is required.",
      },
    ],
    quality_label: "good",
    feedback_score: 4.8,
    outcome: "success",
  });

  const policy = startSpan(root, {
    step_name: "route_policy",
    provider: "tokvera",
    model: "policy-engine",
    span_kind: "guardrail",
    routing_reason: "default_policy",
    route: "openai:gpt-4o-mini",
    decision: {
      routing_reason: "default_policy",
      route: "openai:gpt-4o-mini",
      outcome: "success",
    },
    metrics: {
      cost_usd: 0.00001,
    },
    quality_label: "good",
  });

  finishSpan(policy, {
    response: { selected_route: "openai:gpt-4o-mini" },
    quality_label: "good",
    outcome: "success",
  });
  finishSpan(root, {
    response: { completed: true },
    quality_label: "good",
    feedback_score: 4.9,
    outcome: "success",
  });
}

async function runMistralFeature(baseOptions, feature) {
  const tracer = createTokveraTracer({
    ...baseOptions,
    feature,
    emit_lifecycle_events: true,
    capture_content: true,
  });
  const root = startTrace(tracer.baseOptions, {
    step_name: "mistral_support_flow",
    model: "mistral-router",
    span_kind: "orchestrator",
    quality_label: "poor",
    feedback_score: 2.2,
    outcome: "success",
  });
  const client = trackMistral(createFakeMistralClient(), {
    ...getTrackOptionsFromTraceContext(root, {
      step_name: "mistral_reply",
      span_kind: "model",
      provider: "mistral",
      model: "mistral-small-latest",
      quality_label: "poor",
      feedback_score: 2.1,
      outcome: "success",
      emit_lifecycle_events: true,
      capture_content: true,
    }),
    payload_blocks: [
      {
        payload_type: "context",
        content: "Priority user profile: respond in under three sentences and include the refund policy.",
      },
    ],
  });

  const response = await client.chat.complete({
    model: "mistral-small-latest",
    messages: [{ role: "user", content: "Draft a concise refund policy reply." }],
  });
  finishSpan(root, {
    response: { completed: true, provider_response: response.choices?.[0]?.message?.content ?? null },
    quality_label: "poor",
    feedback_score: 2.2,
    outcome: "success",
  });
}

async function runOpenAIAgentsFeature(baseOptions, feature) {
  const processor = createTokveraOpenAIAgentsTracingProcessor({
    ...baseOptions,
    feature,
    emit_lifecycle_events: true,
    capture_content: true,
  });

  const run = processor.onAgentStart({
    step_name: "support_agent",
    model: "agent-router",
    payload_blocks: [
      {
        payload_type: "context",
        content: "Agent policy: call crm_lookup before drafting any loyalty-related answer.",
      },
    ],
    quality_label: "good",
    feedback_score: 4.6,
    outcome: "success",
  });
  const tool = processor.onToolStart(run, {
    step_name: "crm_lookup",
    tool_name: "crm_lookup",
    input: { customer_id: "cust_100" },
  });
  processor.onToolEnd(tool, {
    response: { customer_tier: "priority" },
    quality_label: "good",
    outcome: "success",
  });
  const model = processor.onModelStart(run, {
    step_name: "draft_reply",
    provider: "openai",
    model: "gpt-4o-mini",
    input: { role: "user", content: "How do loyalty refunds work?" },
    quality_label: "good",
    feedback_score: 4.9,
    outcome: "success",
  });
  processor.onModelEnd(model, {
    response: { output_text: "Priority customers can request refunds within 30 days." },
    usage: { prompt_tokens: 14, completion_tokens: 12, total_tokens: 26 },
    quality_label: "good",
    feedback_score: 5,
    outcome: "success",
  });
  processor.onAgentEnd(run, {
    response: { status: "completed" },
    quality_label: "good",
    feedback_score: 5,
    outcome: "success",
  });
}

async function runLangGraphFeature(baseOptions, feature) {
  const hooks = createTokveraLangGraphHooks({
    ...baseOptions,
    feature,
    emit_lifecycle_events: true,
    capture_content: true,
  });

  const graph = hooks.onGraphStart({
    step_name: "langgraph_support_flow",
    model: "planner-graph",
    quality_label: "good",
    feedback_score: 4.5,
    outcome: "success",
  });
  const planner = hooks.onNodeStart(graph, {
    step_name: "planner",
  });
  hooks.onNodeEnd(planner, {
    response: { next: "kb_lookup" },
    quality_label: "good",
    outcome: "success",
  });
  const branch = hooks.onBranchStart(graph, {
    step_name: "confidence_gate",
    routing_reason: "confidence_below_threshold",
    route: "fallback_path",
  });
  hooks.onBranchEnd(branch, {
    response: { selected_branch: "fallback_path" },
    quality_label: "good",
    feedback_score: 4.4,
    outcome: "success",
  });
  hooks.onGraphEnd(graph, {
    response: { status: "completed" },
    quality_label: "good",
    feedback_score: 4.8,
    outcome: "success",
  });
}

async function runOTelFeature(baseOptions, feature) {
  const exporter = new TokveraOTelSpanExporter({
    ...baseOptions,
    feature,
  });
  const nowSeconds = Math.floor(Date.now() / 1000);
  exporter.export(
    [
      {
        name: "retrieval.plan",
        startTime: [nowSeconds, 0],
        endTime: [nowSeconds, 180_000_000],
        attributes: {
          "tokvera.provider": "openai",
          "tokvera.feature": feature,
          "tokvera.step_name": "retrieval_plan",
          "tokvera.endpoint": "responses.create",
          "tokvera.cost_usd": 0.00002,
          "gen_ai.response.model": "gpt-4o-mini",
          "gen_ai.usage.input_tokens": 7,
          "gen_ai.usage.output_tokens": 4,
        },
        resource: {
          attributes: {
            "service.name": "vector-router",
            "deployment.environment": "test",
          },
        },
        spanContext() {
          return {
            traceId: nextId("trc"),
            spanId: nextId("spn"),
          };
        },
        status: { code: 1 },
      },
    ],
    () => undefined
  );
}

async function main() {
  loadLocalEnv();

  const baseOptions = {
    api_key: process.env.TOKVERA_API_KEY,
    ingest_url: process.env.TOKVERA_INGEST_URL,
    tenant_id: "example_tenant",
    customer_id: "example_customer",
    environment: "test",
    schema_version: "2026-04-01",
  };

  if (!baseOptions.api_key) throw new Error("TOKVERA_API_KEY is required");
  if (!baseOptions.ingest_url) throw new Error("TOKVERA_INGEST_URL is required");

  const features = {
    existingApp: process.env.TOKVERA_FEATURE_EXISTING_APP_JS || "runtime_existing_app_js",
    mistral: process.env.TOKVERA_FEATURE_MISTRAL_JS || "runtime_mistral_js",
    openaiAgents: process.env.TOKVERA_FEATURE_OPENAI_AGENTS_JS || "runtime_openai_agents_js",
    langgraph: process.env.TOKVERA_FEATURE_LANGGRAPH_JS || "runtime_langgraph_js",
    otel: process.env.TOKVERA_FEATURE_OTEL_JS || "runtime_otel_js",
  };

  await runExistingAppFeature(baseOptions, features.existingApp);
  await runMistralFeature(baseOptions, features.mistral);
  await runOpenAIAgentsFeature(baseOptions, features.openaiAgents);
  await runLangGraphFeature(baseOptions, features.langgraph);
  await runOTelFeature(baseOptions, features.otel);

  await new Promise((resolve) => setTimeout(resolve, Number(process.env.TOKVERA_WAIT_MS || 1200)));

  console.log("node runtime helper smoke complete");
  console.log(JSON.stringify({ features }, null, 2));
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
